### PR TITLE
fix HGCAL Layer Cluster times in heterogeneous workflows at HLT 

### PR DESCRIPTION
**Pull request by AuroraPerego**
#### PR description:

The `times` and `timeErrors` vectors used to store the rechits time were initialized with a size of 16, but then `push_back` was used to fill them leaving the first 16 entries as 0.
This PR removes the allocation of 16 elements in the constructor, using `reserve()` instead.
The change affects only heterogeneous workflows at the HLT.

#### PR validation:

Tested on wf 31834.492, the HGCAL Layer Clusters times are expected to change:
image
old -> pre fix
new -> post fix

FYI @rovere

### CODE REVIEW

**Comment by rovere** [comment]
thanks @AuroraPerego for the investigation and the fix.

**Comment by rovere** [comment]
@cmsbuild please test

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClustersFromSoAProducer.cc [hunk 1]
```diff
@@ -87,8 +87,12 @@ class HGCalLayerClustersFromSoAProducer : public edm::stream::EDProducer<> {
...
     // reallocation and the associated overhead when the vector's capacity
     // needs to be extended.
-    std::vector<std::vector<float>> times(clusters->size(), std::vector<float>(16, 0.0f));
-    std::vector<std::vector<float>> timeErrors(clusters->size(), std::vector<float>(16, 0.0f));
+    std::vector<std::vector<float>> times(clusters->size());
+    std::vector<std::vector<float>> timeErrors(clusters->size());
+    for (unsigned int i = 0; i < clusters->size(); ++i) {
+      times[i].reserve(16);
+      timeErrors[i].reserve(16);
+    }
```
Comment (rovere) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClustersFromSoAProducer.cc:95]: Would it make sense to embed this loop into the previous one at lines 69 to 77? In this way, the heuristic is not even necessary and the size can be reserved correctly (i.e. the maximum possible, eventually some rechits will be discarded from the timing computation).

**Comment by rovere** [comment]
@cmsbuild please test

**Comment by rovere** [comment]
The bot seems to be a little stuck...

**Comment by mmusich** [comment]
> Tested on wf 31834.492, the HGCAL Layer Clusters times are expected to change:

is there a way to make this change appear in the bot tests? E.g. by running wf `31834.492` ?

**Comment by AuroraPerego** [comment]
> is there a way to make this change appear in the bot tests? E.g. by running wf `31834.492` ?

I'm unsure because this workflow runs the producer I've changed, but the collection of clusters times is not saved in the event.

**Comment by mmusich** [comment]
> I'm unsure because this workflow runs the producer I've changed, but the collection of clusters times is not saved in the event.

and does it have any bearing on trigger decisions?

**Comment by AuroraPerego** [comment]
> and does it have any bearing on trigger decisions?

I doubt :)

**Comment by mmusich** [comment]
> I doubt :)

I am confused. So this PR is of no consequence? what's the purpose?

**Comment by rovere** [comment]
when running CLUE on GPU at HLT for Phase2, the timing assigned to the clusters is wrong.
Regardless of its usage, we need to fix it.
This PR fixes it.

**Comment by AuroraPerego** [comment]
> is there a way to make this change appear in the bot tests? E.g. by running wf 31834.492 ?

Thinking about it the effects should be visible downstream, e.g. in tracksters and candidates. They are saved in the event and their time is computed using the clusters times.

**Comment by rovere** [comment]
test parameters:

- enable = gpu
- workflows_gpu = 31834.492

**Comment by rovere** [comment]
please test

**Comment by rovere** [comment]
test parameters:
- enable = gpu
- workflows_gpu = 31834.492
- workflow_opts = -w upgrade
- workflow_opts_gpu = -w upgrade

**Comment by mmusich** [comment]
> +1

tbh, I am more confused than ever.
In wf `31834.492_TTbar_14TeV+2026D110SimOnGen_Patatrack_FullRecoAlpaka` I see plenty of HLT changes [log](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_2_X_2024-09-04-1100+4551d6/64336/31834.492_TTbar_14TeV+2026D110SimOnGen_Patatrack_FullRecoAlpaka/): all of them in tracking (in which I would have not anticipated any changes, and none in HGCal, where I would have expected some  according to https://github.com/cms-sw/cmssw/pull/45838#issuecomment-2325080759).
By no means I am a reviewer but I would be keen to understand better.

**Comment by felicepantaleo** [comment]
> > +1
> 
> tbh, I am more confused than ever. In wf `31834.492_TTbar_14TeV+2026D110SimOnGen_Patatrack_FullRecoAlpaka` I see plenty of HLT changes [log](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_2_X_2024-09-04-1100+4551d6/64336/31834.492_TTbar_14TeV+2026D110SimOnGen_Patatrack_FullRecoAlpaka/): all of them in tracking (in which I would have not anticipated any changes) and none in HGCal (where I would have expected some according to [#45838 (comment)](https://github.com/cms-sw/cmssw/pull/45838#issuecomment-2325080759)). By no means I am a reviewer but I would be keen to understand better.

The comparisons show tiny difference in pixel tracks offline alpaka GPU vs alpaka GPU if I understand correctly.
Are you surprised because you would have expected phase-2 Pixel Tracks alpaka running on GPU to be bit-by-bit reproducible?

**Comment by Moanwar** [comment]
please test

**Comment by mmusich** [comment]
> Are you surprised because you would have expected phase-2 Pixel Tracks alpaka running on GPU to be bit-by-bit reproducible?

I see also changes in HLT tracking https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_2_X_2024-09-04-1100+4551d6/64336/31834.492_TTbar_14TeV+2026D110SimOnGen_Patatrack_FullRecoAlpaka/HLT_Tracking_pixelTracks.html

**Comment by felicepantaleo** [comment]
test parameters:

- enable = gpu
- workflows_gpu = 31834.492
- workflow_opts = -w upgrade
- workflow_opts_gpu = -w upgrade

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by felicepantaleo** [comment]
test parameters:

- enable = gpu
- workflows_gpu = 31834.496
- workflow_opts = -w upgrade
- workflow_opts_gpu = -w upgrade

**Comment by felicepantaleo** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by Moanwar** [comment]
+Upgrade

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**